openeo/internal/jupyter.py

```diff
@@ -0,0 +1,70 @@
+from string import Template
+from collections import Mapping
+import json
+
+# These classes are proxies to visualize openEO responses nicely in Jupyter
+# To show the actual list or dict in Jupyter, use repr() or print()
+
+SCRIPT_URL = 'https://cdn.jsdelivr.net/npm/@openeo/vue-components@2.0.0-beta.1/assets/openeo.js'
+
+class JupyterIntegration():
+
+	# Class instance to store whether the script has been inserted into the DOM
+	# TODO: This fails if a page is reloaded in Jupyter. The variable is still True, but the script doesn't exist in the page any longer.
+	script_sent = False
+
+	def __init__(self, component: str, parameters: dict = {}):
+		self.component = component
+		self.parameters = parameters
+
+	def _repr_html_(self):
+		# Output script tag only once, otherwise JS throws errors
+		script = ''
+		if JupyterIntegration.script_sent == False:
+			script = '<script src="{}"></script>'.format(SCRIPT_URL)
+			JupyterIntegration.script_sent = True
+
+		# Construct HTML
+		template = Template("""
+		$script
+		<openeo-$component>
+			<script type="application/json">$props</script>
+		</openeo-$component>
+		""")
+		return template.substitute(
+			script = script,
+			component = self.component,
+			props = json.dumps(self.parameters)
+		)
+
+
+class VisualDict(dict, JupyterIntegration):
+
+	# The first entry of the dict is expected to always be the actual dict with the data
+	def __init__(self, component: str, parameters: dict = {}):
+		JupyterIntegration.__init__(self, component, parameters)
+
+		data = {}
+		if len(parameters) > 0:
+			i = next(iter(parameters))
+			if isinstance(parameters[i], Mapping):
+				data = parameters[i]
```

Comment:
why are you only checking/using the first item of the `parameters` dictionary? Note that dictionary ordering is not guaranteed in older python versions

Reply:
From the usage below I see you you pass always a single item dictionary to the constructor here, and you are ignoring its key (`i` above) in the end. And there also seems to be redundancy between this key and `component`.

Reply:
I would also make what's called `parameters` here as first argument, because that is the main argument of the parent class (`dict` in this case)

Reply:
> why are you only checking/using the first item of the `parameters` dictionary?

The parameters are the parameters that are passed to the Vue components (e.g. parameters `capabilities` and `url` in Capabilities component). To figure out which is the actual dict or list to proxy in the VisualDict/List, I need to have a convention which parameter it is (the name is always different). Thus I thought the convention is it's always the first...

> Note that dictionary ordering is not guaranteed in older python versions

From what I found on the web, order is guaranteed for Python >= 3.6 and the client readme says "Requirement: a Python 3.6 (or higher) environment" so I think it's fine?!

> From the usage below I see you you pass always a single item dictionary

No, there are cases with more than one (Capabilities) and there will be more in the future.

> and you are ignoring its key (i above) in the end

No, it's used in the json.dumps, which transmits the keys and values to the component, which is important.

> I would also make what's called parameters here as first argument, 

I need to know the key to pass it to the Vue component. So either we go `__init__(component : str, key : str, data : dict/list, parameters :dict = {})` then or leave it as it is.

Reply:
I also experimented with **kwargs, but can't remember why this did not work... 🤔

Reply:
I think item order it's only guaranteed from py 3.7 and just a undocumented side-effect of the implementation in 3.6. But regardless of that, it looks bad style to depend on dict item order to my python eyes.

Wouldn't it be easier to just do something like this:
```python

class JupyterIntegration:
    def __init__(self, component, data, **kwargs):
        self.component = component
        self.data = data
        self.parameters = kwargs

    def _repr_html_(self):
        ...
        return " ....".format(
             ....
             props = json.dump({"data": self.data, "parameters": self.parameters})

class VisualDict(dict, JupyterIntegration):
    def __init__(self, data, component, **kwargs):
        dict.__init__(data)
        JupyterIntegration.__init__(component=component, data=data, **kwargs)

```

Reply:
As I said, "data" is not a fixed term. It can be "formats", "collections", ... so I need to pass the key name, too.

Reply:
No, there's not. Either I need to provide it in the constructor or define a mapping with a couple of if/elifs.

Reply:
Can't you move that logic to the JS side? then you have a smaller API surface between python and JS: just `component` instead of `component` + `key`?

Reply:
No, unfortunately not. Actually, this is HTML here.

Reply:
I've moved the "key" handling into the Jupyter code now, it at least doesn't show up in the "normal" functions now... Is that okay? @soxofaan